core/internal/filetransfer/file_transfer_azure.go

```diff
@@ -0,0 +1,416 @@
+	ft.logger.Debug(
+		"Azure file transfer: downloading file",
+		"path", task.PathOrPrefix,
+		"ref", task.Reference,
+	)
+
+	// Parse the reference path to get the account URL and blob path
+	accountUrl, fullBlobPath, err := parseCloudReference(task.Reference, azureScheme)
+	if err != nil {
+		return ft.formatDownloadError("error parsing reference", err)
+	}
+	pathSplit := strings.SplitN(fullBlobPath, "/", 2)
+	fullAccountUrl := fmt.Sprintf("%s://%s", azureScheme, accountUrl)
+	blobInfo := ParsedBlobInfo{
+		AccountUrl: fullAccountUrl,
+		Container:  pathSplit[0],
+		BlobPrefix: pathSplit[1],
+	}
+
+	// Setup the client if it is not already set up
+	ft.SetupClient(fullAccountUrl)
+	_, ok := ft.clients.clients.Load(fullAccountUrl)
+	if !ok {
+		return ft.formatDownloadError(
+			"error setting up Azure account client",
+			fmt.Errorf("client not found"),
+		)
+	}
+
+	var blobNames []string
+	if task.HasSingleFile() {
+		blobName, versionId, err := ft.getBlobName(blobInfo, task)
+		if err != nil {
+			return ft.formatDownloadError(
+				"error getting correct blob name and version",
+				err,
+			)
+		}
+		if versionId != "" {
+			err = task.SetVersionID(versionId)
+			if err != nil {
+				return ft.formatDownloadError(
+					"error setting version ID",
+					err,
+				)
+			}
+		}
+		blobNames = []string{blobName}
+	} else {
+		blobNames, err = ft.listBlobsWithPrefix(blobInfo)
+		if err != nil {
+			return ft.formatDownloadError(
+				fmt.Sprintf("error finding blobs with prefix %s", blobInfo.BlobPrefix),
+				err,
+			)
+		}
+	}
+
+	err = ft.downloadFiles(blobInfo, blobNames, task)
+	if err != nil {
+		return ft.formatDownloadError(
+			fmt.Sprintf("error downloading reference %s", task.Reference),
+			err,
+		)
+	}
+	return nil
+}
+
+// getBlobName tries to get the blob name and version ID that matches the
+// expected digest for the given task.
+func (ft *AzureFileTransfer) getBlobName(
+	blobInfo ParsedBlobInfo,
+	task *ReferenceArtifactDownloadTask,
+) (string, string, error) {
+	blobClient, err := ft.SetupBlobClient(task)
+	if err != nil {
+		return "", "", err
+	}
+	properties, err := blobClient.GetProperties(ft.ctx, nil)
+	if err != nil {
+		return "", "", err
+	}
+
+	// If the etag does not match the expected digest, try to find the correct blob version
+	if properties.ETag != nil && strings.Trim(string(*properties.ETag), "\"") != task.Digest {
+		if task.VersionId != nil {
+			return "", "", fmt.Errorf(
+				"digest/etag mismatch: etag %s does not match expected digest %s",
+				*properties.ETag,
+				task.Digest,
+			)
+		}
+		blobName, versionId, err := ft.getCorrectBlobVersion(blobInfo, task)
+		if err != nil {
+			return "", "", err
+		}
+		return blobName, versionId, nil
+	}
+	return blobInfo.BlobPrefix, "", nil
+}
+
+// getCorrectBlobVersion finds the correct blob version that matches the
+// expected digest.
+func (ft *AzureFileTransfer) getCorrectBlobVersion(
```

Comment:
again i think you can split this function differently, the logic that get the blob version could be a separate function and in the call site all you need is to get the relevant client and pass it as argument to the version check function.

Reply:
I tried to make this logic easier to follow by adding a function to check if the version matches and adding some comments. Since we have to check the etag/digest matching for every version we get its still a bit nested, but I think this should make it easier to understand